cajita/src/Cajita_FastFourierTransform.hpp

```diff
@@ -23,291 +23,509 @@
+            throw std::logic_error(
+                "Only 1 complex value per entity allowed in FFT" );
+    }
 
-      If this function is not used to set the type then 2 is used as the
-      default.
+    /*!
+      \brief Do a forward FFT.
+      \param x The array on which to perform the forward transform.
+      \param scaling Method of scaling data.
     */
-    FastFourierTransformParams& setPackType( const int type )
+    template <class Array_t, class ScaleType>
+    void forward( const Array_t& x, const ScaleType scaling )
     {
-        packflag = type;
-        return *this;
+        checkArray( x );
+        static_cast<InterfaceType*>( this )->forwardImpl( x, scaling );
     }
 
     /*!
-      \brief Set the scaling type.
-      \param type Scaling type.
-
-      0: Not scaling after forward
-      1: Scaling after forward
-
-      If this function is not used to set the type then 1 is used as the
-      default.
+      \brief Do a reverse FFT.
+      \param x The array on which to perform the reverse transform.
+      \param scaling Method of scaling data.
     */
-    FastFourierTransformParams& setScalingType( const int type )
+    template <class Array_t, class ScaleType>
+    void reverse( const Array_t& x, const ScaleType scaling )
     {
-        scaled = type;
-        return *this;
+        checkArray( x );
+        static_cast<InterfaceType*>( this )->reverseImpl( x, scaling );
     }
+};
 
-    // Collective communication type.
-    int collective;
-
-    // Parallel decomposition exchange type.
-    int exchange;
+//---------------------------------------------------------------------------//
+// heFFTe
+//---------------------------------------------------------------------------//
 
-    // Buffer packing type.
-    int packflag;
+namespace Impl
+{
+template <class ExecutionSpace, class Scalar, class BackendType>
+struct HeffteBackendTraits
+{
+};
+#ifdef Heffte_ENABLE_FFTW
+template <class ExecutionSpace, class Scalar>
+struct HeffteBackendTraits<ExecutionSpace, Scalar, FFTBackendFFTW>
+{
+    using backend_type = heffte::backend::fftw;
+    using complex_type = std::complex<Scalar>;
+};
+template <class ExecutionSpace, class Scalar>
+struct HeffteBackendTraits<ExecutionSpace, Scalar, Impl::FFTBackendDefault>
+{
+    using backend_type = heffte::backend::fftw;
+    using complex_type = std::complex<Scalar>;
+};
+#endif
+#ifdef Heffte_ENABLE_MKL
+template <class ExecutionSpace, class Scalar>
+struct HeffteBackendTraits<ExecutionSpace, Scalar, FFTBackendMKL>
+{
+    using backend_type = heffte::backend::mkl;
+    using complex_type = std::complex<Scalar>;
+};
+#endif
+#ifdef Heffte_ENABLE_CUDA
+#ifdef KOKKOS_ENABLE_CUDA
+template <>
+struct HeffteBackendTraits<Kokkos::Cuda, double, Impl::FFTBackendDefault>
+{
+    using backend_type = heffte::backend::cufft;
+    using complex_type = cufftDoubleComplex;
+};
+template <>
+struct HeffteBackendTraits<Kokkos::Cuda, float, Impl::FFTBackendDefault>
```

Comment:
Here is where I would write a HIP specialization and just put `static_assert( false, "HIP not supported" );`